ql/src/java/org/apache/hadoop/hive/ql/exec/tez/monitoring/TezJobMonitor.java

```diff
@@ -0,0 +1,395 @@
+import org.apache.tez.common.counters.TezCounters;
+import org.apache.tez.dag.api.DAG;
+import org.apache.tez.dag.api.TezException;
+import org.apache.tez.dag.api.client.DAGClient;
+import org.apache.tez.dag.api.client.DAGStatus;
+import org.apache.tez.dag.api.client.Progress;
+import org.apache.tez.dag.api.client.StatusGetOpts;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.io.StringWriter;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static org.apache.tez.dag.api.client.DAGStatus.State.RUNNING;
+
+/**
+ * TezJobMonitor keeps track of a tez job while it's being executed. It will
+ * print status to the console and retrieve final status of the job after
+ * completion.
+ */
+public class TezJobMonitor {
+
+  private static final String CLASS_NAME = TezJobMonitor.class.getName();
+  private static final int CHECK_INTERVAL = 200;
+  private static final int MAX_RETRY_INTERVAL = 2500;
+  private static final int PRINT_INTERVAL = 3000;
+
+  private final PerfLogger perfLogger = SessionState.getPerfLogger();
+  private static final List<DAGClient> shutdownList;
+  private final Map<String, BaseWork> workMap;
+
+  private transient LogHelper console;
+
+  private long lastPrintTime;
+  private StringWriter diagnostics = new StringWriter();
+
+  interface UpdateFunction {
+    void update(DAGStatus status, Map<String, Progress> vertexProgressMap, String report);
+  }
+
+  static {
+    shutdownList = new LinkedList<>();
+    ShutdownHookManager.addShutdownHook(new Runnable() {
+      @Override
+      public void run() {
+        TezJobMonitor.killRunningJobs();
+        try {
+          TezSessionPoolManager.getInstance().closeNonDefaultSessions(false);
+        } catch (Exception e) {
+          // ignore
+        }
+      }
+    });
+  }
+
+  public static void initShutdownHook() {
+    Preconditions.checkNotNull(shutdownList,
+      "Shutdown hook was not properly initialized");
+  }
+
+  private final DAGClient dagClient;
+  private final HiveConf hiveConf;
+  private final DAG dag;
+  private final Context context;
+  private long executionStartTime = 0;
+  private final UpdateFunction updateFunction;
+  /**
+   * Have to use the same instance to render else the number lines printed earlier is lost and the
+   * screen will print the table again and again.
+   */
+  private final InPlaceUpdate inPlaceUpdate;
+
+  public TezJobMonitor(Map<String, BaseWork> workMap, final DAGClient dagClient, HiveConf conf, DAG dag,
+                       Context ctx) {
+    this.workMap = workMap;
+    this.dagClient = dagClient;
+    this.hiveConf = conf;
+    this.dag = dag;
+    this.context = ctx;
+    console = SessionState.getConsole();
+    inPlaceUpdate = new InPlaceUpdate(LogHelper.getInfoStream());
+    updateFunction = updateFunction();
+  }
+
+  private UpdateFunction updateFunction() {
```

Comment:
personally for me the function signature gives a good indication of the intention that its trying to do, so adding the extra "get" / "set" seems unnecessary, however the current convention is according to JavaBeans and i will change this to follow that